### DISCUSSION TITLE
Problem with `cargo-vendor-deps`

### DISCUSSION THREAD

**Discussion by eureka-cpu** [ipetkov/crane#711] [Q&A] [answered]
Hello, I'm working off of this branch: https://github.com/anagrambuild/bonsol/blob/da9cbe6ed248b1c159cf10542d4d59f1986f626e/flake.nix

And running into this error:
```
       > source root is source
       > Running phase: patchPhase
       > Executing configureCargoCommonVars
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > will append /build/source/.cargo-home/config.toml with contents of /nix/store/932zznc62fgngqb7bz1qlnscmh34ll7l-vendor-cargo-deps/config.toml
       > default configurePhase, nothing to do
       > Running phase: buildPhase
       > ++ command cargo --version
       > cargo 1.81.0 (2dbb1af80 2024-08-20)
       > ++ command cargo check --release --locked --all-targets
       > error: failed to get `yellowstone-grpc-client` as a dependency of package `relay v0.1.9 (/build/source/relay)`
       >
       > Caused by:
       >   failed to load source for dependency `yellowstone-grpc-client`
       >
       > Caused by:
       >   Unable to update https://github.com/rpcpool/yellowstone-grpc?tag=v1.15.0+solana.1.18.22
       >
       > Caused by:
       >   the source https://github.com/rpcpool/yellowstone-grpc?tag=v1.15.0+solana.1.18.22 requires a lock file to be present first before it can be
       >   used against vendored source code
       >
       >   remove the source replacement configuration, generate a lock file, and then
       >   restore the source replacement configuration to continue the build
       For full logs, run 'nix log /nix/store/wc1bkkcisyb9j92043manf24db56dq11-bonsol-deps-0.0.1.drv'.
```

I'm not sure I understand, and can't seem to find a good answer online about how to solve this. I have a `Cargo.lock` in the root that I generated again after I first got the error, and the dependency it is complaining about is indeed there. Would anyone mind pointing me in the right direction?

**Comment by ipetkov** [comment, marked as answer]
Hi @eureka-cpu ! The first thing that comes to mind is run `cargo vendor` then follow its instructions (e.g. paste some configs in `.cargo/config.toml`) and try to build your project outside of Nix (e.g. a vanilla `cargo build ...`). If everything builds fine that way, then there might be a vendoring bug in `crane` (in which case open an issue and I'll check it out).

Otherwise if **you get the same error after running `cargo vendor`**, then it is possible one of your git dependencies is not vendor-friendly (sadly the way cargo handles git deps by default _happens_ to allow certain project configurations to work but fail when their sources are vendored). If that's the case, the only remedy would be to fix the issue upstream or [patch the sources yourself](https://github.com/ipetkov/crane/blob/a376dd1efac7bce448857c62961c6311be26cb09/docs/patching_dependency_sources.md)

Hope this helps!

**Reply by eureka-cpu** [reply]
Thanks for the quick reply! I changed the git dependency to one from crates.io and it seems to have gotten past the vendoring issue. I'll mark as answered for anyone else who runs into this problem.
